File: mc-mod/src/main/java/world/oyster/recorder/ZBufferCapture.java
package world.oyster.recorder;

import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gl.Framebuffer;
import org.lwjgl.opengl.GL11;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * ZBufferCapture — per-frame GL depth-buffer readback for the Oyster Recorder mod.
 *
 * <p>Companion to {@link GameStateCapture}: where that streams player pose to
 * {@code game_state.jsonl}, this streams the raw GL depth buffer to
 * {@code zbuffer/tick_<N>.bin} so the downstream (D2) pipeline can build
 * metric depth / EXR. Per Howard's directive depth is RAW data: we write the
 * unmodified GL depth-component values ([0,1], non-linear) and linearise to
 * world-space metres in post-processing (where near/far are known from
 * metadata) rather than guessing version-specific camera-projection APIs here.
 *
 * <p>Capture is throttled to {@code OYSTER_ZBUFFER_FPS} (default 6) to align
 * with the 6-FPS video/game_state cadence (prd_test_depth_6fps_alignment) and
 * to keep disk throughput sane (a 1080p depth frame is ~8 MB raw).
 *
 * <p>Output co-locates with game_state via {@link SessionDir}: the
 * {@code zbuffer/} dir is created next to {@code game_state.jsonl} (honouring
 * {@code OYSTER_SESSION_DIR} when the launcher sets it). A sidecar
 * {@code zbuffer/index.jsonl} records {tick, t_ms, w, h} per frame for
 * downstream temporal alignment.
 *
 * <p>Fail-soft: any error logs and skips — never crashes the user's game.
 * Disable entirely with {@code OYSTER_ZBUFFER_CAPTURE=0}.
 */
public final class ZBufferCapture {

    private static final Logger LOGGER = LoggerFactory.getLogger("oyster-recorder/zbuffer");

    /** "0" disables; any other value (or unset) leaves capture ON for this depth build. */
    private static final String ENV_ENABLE = "OYSTER_ZBUFFER_CAPTURE";
    /** Capture rate in frames/sec; default 6 to match the video/game_state alignment cadence. */
    private static final String ENV_FPS = "OYSTER_ZBUFFER_FPS";
    /** Bounded async write queue — frames are dropped (logged) rather than stalling the render thread. */
    private static final int MAX_PENDING = 16;

    private static final AtomicInteger tickCounter = new AtomicInteger(0);
    private static final AtomicInteger pendingCount = new AtomicInteger(0);
    private static final Object INDEX_LOCK = new Object();

    private static volatile long intervalMs = 167L; // ~6 FPS
    private static volatile long lastCaptureMs = 0L;

    // Reusable direct buffer for glReadPixels (touched on render thread only).
    private static ByteBuffer readBuf = null;
    private static int readBufCapacity = 0;

    private ZBufferCapture() {}

    /** Registers the per-frame depth hook. Called from OysterRecorderMod.onInitializeClient(). */
    public static void register() {
        if ("0".equals(System.getenv(ENV_ENABLE))) {
            LOGGER.info("[ZBufferCapture] disabled (OYSTER_ZBUFFER_CAPTURE=0)");
            return;
        }
        int fps = 6;
        try {
            String f = System.getenv(ENV_FPS);
            if (f != null && !f.trim().isEmpty()) {
                fps = Math.max(1, Math.min(60, Integer.parseInt(f.trim())));
            }
        } catch (NumberFormatException ignored) {
            // keep default
        }
        intervalMs = Math.max(1L, 1000L / fps);

        LOGGER.info("[ZBufferCapture] enabled — {} FPS, raw f32 depth → {}", fps, zbufferDir());

        WorldRenderEvents.AFTER_TRANSLUCENT.register(context -> {
            try {
                maybeCapture();
            } catch (Throwable t) {
                LOGGER.warn("[ZBufferCapture] frame capture failed (non-fatal): {}", t.toString());
            }
        });
    }

    private static void maybeCapture() {
        long now = System.currentTimeMillis();
        if (now - lastCaptureMs < intervalMs) return;

        MinecraftClient client = MinecraftClient.getInstance();
        if (client == null || client.world == null || client.player == null) return; // only in-world
        Framebuffer fb = client.getFramebuffer();
        if (fb == null) return;
        int width = fb.textureWidth;
        int height = fb.textureHeight;
        if (width <= 0 || height <= 0) return;

        lastCaptureMs = now;

        int floats = width * height;
        int bytes = floats * 4;
        if (readBuf == null || readBufCapacity < bytes) {
            readBuf = ByteBuffer.allocateDirect(bytes).order(ByteOrder.LITTLE_ENDIAN);
            readBufCapacity = bytes;
        }
        readBuf.clear();

        // Raw depth component of the currently-bound (world) framebuffer.
        GL11.glReadPixels(0, 0, width, height, GL11.GL_DEPTH_COMPONENT, GL11.GL_FLOAT, readBuf);

        // Copy into a heap float[] with vertical flip (GL origin is bottom-left → top-left).
        FloatBuffer src = readBuf.asFloatBuffer();
        float[] depth = new float[floats];
        for (int y = 0; y < height; y++) {
            int srcRow = (height - 1 - y) * width;
            int dstRow = y * width;
            for (int x = 0; x < width; x++) {
                depth[dstRow + x] = src.get(srcRow + x);
            }
        }

        int tickId = tickCounter.incrementAndGet();
        enqueueWrite(tickId, width, height, now, depth);
    }

    private static void enqueueWrite(int tickId, int width, int height, long tMs, float[] depth) {
        if (pendingCount.get() >= MAX_PENDING) {
            LOGGER.warn("[ZBufferCapture] write queue full ({}), dropping tick {}", MAX_PENDING, tickId);
            return;
        }
        pendingCount.incrementAndGet();
        CompletableFuture.runAsync(() -> {
            try {
                writeTick(tickId, width, height, tMs, depth);
            } catch (Throwable t) {
                LOGGER.warn("[ZBufferCapture] async write failed for tick {}: {}", tickId, t.toString());
            } finally {
                pendingCount.decrementAndGet();
            }
        });
    }

    private static void writeTick(int tickId, int width, int height, long tMs, float[] depth) throws IOException {
        Path dir = zbufferDir();
        Files.createDirectories(dir);
        Path out = dir.resolve("tick_" + tickId + ".bin");

        ByteBuffer buf = ByteBuffer.allocate(12 + depth.length * 4).order(ByteOrder.LITTLE_ENDIAN);
        buf.putInt(width);
        buf.putInt(height);
        buf.putInt(tickId);
        for (float v : depth) buf.putFloat(v);
        Files.write(out, buf.array(),
            StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);

        // Sidecar alignment index (one JSON object per line).
        String line = "{\"tick\":" + tickId + ",\"t_ms\":" + tMs
            + ",\"w\":" + width + ",\"h\":" + height
            + ",\"file\":\"tick_" + tickId + ".bin\",\"format\":\"raw_gl_depth_f32le\"}\n";
        synchronized (INDEX_LOCK) {
            Files.write(dir.resolve("index.jsonl"), line.getBytes(StandardCharsets.UTF_8),
                StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.APPEND);
        }
    }

    /** zbuffer/ dir co-located with game_state.jsonl (honours OYSTER_SESSION_DIR). */
    private static Path zbufferDir() {
        Path gs = SessionDir.outputPath();          // .../game_state.jsonl
        Path parent = gs.getParent();
        if (parent == null) parent = gs;
        return parent.resolve("zbuffer");
    }
}

File: patches/cluster-week1-2026-05-18/D1-mc-mod/src/main/kotlin/ai/oyster/recorder/mod/ZBufferCapture.kt
package ai.oyster.recorder.mod

import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents
import net.minecraft.client.MinecraftClient
import org.lwjgl.opengl.GL11
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.IntBuffer
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicInteger

/**
 * ZBufferCapture — captures GL depth buffer each frame, linearizes to world-space meters,
 * and writes raw f32 LE binary files for downstream EXR conversion (D2 Python pipeline).
 *
 * Controlled by env var OYSTER_ZBUFFER_CAPTURE=1 (default off to avoid impacting existing users).
 *
 * Output format per tick file:
 *   Header (12 bytes): u32 width (LE) | u32 height (LE) | u32 tick_id (LE)
 *   Payload: width × height × 4 bytes of f32 LE depth values (meters)
 *
 * Async write via CompletableFuture with bounded queue (max 60 pending).
 * Oldest entries are dropped when queue is full.
 */
object ZBufferCapture : ClientModInitializer {

    private val LOGGER = LoggerFactory.getLogger(ZBufferCapture::class.java)

    private const val ENV_VAR = "OYSTER_ZBUFFER_CAPTURE"
    private const val MAX_PENDING = 60

    // Bounded queue for async write tasks
    private val pendingWrites = ConcurrentLinkedQueue<PendingWrite>()
    private val pendingCount = AtomicInteger(0)

    // Tick counter
    private val tickCounter = AtomicInteger(0)

    // Reusable depth buffer (allocated once, resized if needed)
    @Volatile
    private var depthBuffer: FloatArray? = null
    @Volatile
    private var lastWidth = 0
    @Volatile
    private var lastHeight = 0

    /**
     * Data class holding everything needed for an async write.
     */
    private data class PendingWrite(
        val tickId: Int,
        val width: Int,
        val height: Int,
        val depthValues: FloatArray,
    )

    /**
     * Entry point called by Fabric Loader via fabric.mod.json client entrypoint.
     */
    @JvmStatic
    fun init() {
        val instance = ZBufferCapture
        instance.onInitializeClient()
    }

    override fun onInitializeClient() {
        val enabled = System.getenv(ENV_VAR) == "1"
        if (!enabled) {
            LOGGER.info("[ZBufferCapture] Disabled (set $ENV_VAR=1 to enable)")
            return
        }

        LOGGER.info("[ZBufferCapture] Enabled — hooking WorldRenderEvents.AFTER_TRANSLUCENT")

        WorldRenderEvents.AFTER_TRANSLUCENT.register { context ->
            try {
                captureFrame(context)
            } catch (e: Exception) {
                LOGGER.warn("[ZBufferCapture] Frame capture failed: ${e.message}", e)
            }
        }
    }

    /**
     * Main capture logic called each frame from the render thread.
     * Reads depth buffer, linearizes, queues async write.
     */
    private fun captureFrame(context: WorldRenderEvents.Context) {
        val client = MinecraftClient.getInstance()
        val camera = client.gameRenderer.camera

        if (camera == null) {
            return
        }

        val framebuffer = client.framebuffer
        if (framebuffer == null) {
            return
        }

        val width = framebuffer.textureWidth
        val height = framebuffer.textureHeight

        if (width <= 0 || height <= 0) {
            return
        }

        // Ensure depth buffer is allocated
        ensureDepthBuffer(width, height)

        val buf = depthBuffer!!

        // Read depth component as float from the current framebuffer
        // GL_DEPTH_COMPONENT + GL_FLOAT reads the depth buffer into a float array
        readDepthBuffer(width, height, buf)

        // Linearize depth values from [0,1] NDC to world-space meters
        val nearPlane = camera.clipStart.toDouble()
        val farPlane = camera.clipEnd.toDouble()

        if (nearPlane <= 0.0 || farPlane <= nearPlane) {
            LOGGER.debug("[ZBufferCapture] Invalid clip planes: near=$nearPlane far=$farPlane, skipping")
            return
        }

        linearizeDepth(buf, nearPlane, farPlane)

        // Queue async write
        val tickId = tickCounter.incrementAndGet()
        val writeData = PendingWrite(tickId, width, height, buf.copyOf())
        queueWrite(writeData)
    }

    /**
     * Ensures the reusable depth buffer is large enough for the given dimensions.
     */
    private fun ensureDepthBuffer(width: Int, height: Int) {
        val needed = width * height
        val current = depthBuffer
        if (current == null || current.size < needed || lastWidth != width || lastHeight != height) {
            depthBuffer = FloatArray(needed)
            lastWidth = width
            lastHeight = height
            LOGGER.debug("[ZBufferCapture] Allocated depth buffer: ${width}x${height} (${needed} floats)")
        }
    }

    /**
     * Reads the depth buffer from OpenGL into the provided FloatArray.
     * Uses GL11.glReadPixels with GL_DEPTH_COMPONENT and GL_FLOAT.
     */
    private fun readDepthBuffer(width: Int, height: Int, out: FloatArray) {
        // Allocate a direct ByteBuffer for the read
        val byteSize = width * height * 4 // 4 bytes per float
        val byteBuffer = ByteBuffer.allocateDirect(byteSize).order(ByteOrder.LITTLE_ENDIAN)
        val floatBuffer = byteBuffer.asFloatBuffer()

        // glReadPixels reads from bottom-left, so we need to flip vertically
        // Read into the buffer
        GL11.glReadPixels(0, 0, width, height, GL11.GL_DEPTH_COMPONENT, GL11.GL_FLOAT, byteBuffer)

        // Copy to our array, flipping Y axis (OpenGL origin is bottom-left, we want top-left)
        for (y in 0 until height) {
            val srcRow = (height - 1 - y) * width
            val dstRow = y * width
            for (x in 0 until width) {
                out[dstRow + x] = floatBuffer.get(srcRow + x)
            }
        }
    }

    /**
     * Linearizes depth buffer values from [0,1] range to world-space meters.
     *
     * Minecraft uses a perspective projection. The depth buffer value z_buf is
     * non-linear due to perspective division. We reverse the projection:
     *
     *   z_ndc = 2.0 * z_buf - 1.0          (convert [0,1] to [-1,1])
     *   z_eye = (2 * near * far) / (far + near - z_ndc * (far - near))
     *   z_world = |z_eye|                    (distance from camera in meters)
     *
     * This is the standard OpenGL depth linearization formula.
     */
    fun linearizeDepth(depthValues: FloatArray, near: Double, far: Double) {
        val nearF = near.toFloat()
        val farF = far.toFloat()
        val twoNearFar = 2.0f * nearF * farF
        val farPlusNear = farF + nearF
        val farMinusNear = farF - nearF

        for (i in depthValues.indices) {
            val zBuf = depthValues[i]
            // Convert from [0,1] to [-1,1] NDC
            val zNdc = 2.0f * zBuf - 1.0f
            // Reverse perspective projection to get eye-space Z
            val zEye = twoNearFar / (farPlusNear - zNdc * farMinusNear)
            // World-space distance (absolute value, since eye-space Z is negative in OpenGL)
            depthValues[i] = kotlin.math.abs(zEye)
        }
    }

    /**
     * Queues a write task. If the queue is full (>= MAX_PENDING), drops the oldest entry.
     */
    private fun queueWrite(data: PendingWrite) {
        // Drop oldest if queue is full
        while (pendingCount.get() >= MAX_PENDING) {
            val dropped = pendingWrites.poll()
            if (dropped != null) {
                pendingCount.decrementAndGet()
                LOGGER.warn("[ZBufferCapture] Queue full, dropping oldest tick ${dropped.tickId}")
            } else {
                break
            }
        }

        pendingWrites.offer(data)
        pendingCount.incrementAndGet()

        // Fire async write
        CompletableFuture.runAsync {
            try {
                writeToFile(data)
            } catch (e: Exception) {
                LOGGER.error("[ZBufferCapture] Async write failed for tick ${data.tickId}: ${e.message}", e)
            } finally {
                pendingWrites.remove(data)
                pendingCount.decrementAndGet()
            }
        }
    }

    /**
     * Writes the depth data to disk in the specified binary format.
     *
     * Path: ~/Documents/OysterClips/active_session/zbuffer/tick_<N>.bin
     *
     * Format:
     *   Bytes 0-3:   u32 width (little-endian)
     *   Bytes 4-7:   u32 height (little-endian)
     *   Bytes 8-11:  u32 tick_id (little-endian)
     *   Bytes 12+:   raw f32 LE depth values (width × height × 4 bytes)
     */
    private fun writeToFile(data: PendingWrite) {
        val outputDir = getOutputDirectory()
        outputDir.mkdirs()

        val outputFile = File(outputDir, "tick_${data.tickId}.bin")

        FileOutputStream(outputFile).use { fos ->
            val header = ByteBuffer.allocate(12).order(ByteOrder.LITTLE_ENDIAN)
            header.putInt(data.width)
            header.putInt(data.height)
            header.putInt(data.tickId)
            header.flip()
            fos.write(header.array())

            // Write depth values as little-endian f32
            val byteBuffer = ByteBuffer.allocate(data.depthValues.size * 4).order(ByteOrder.LITTLE_ENDIAN)
            byteBuffer.asFloatBuffer().put(data.depthValues)
            fos.write(byteBuffer.array())
        }

        LOGGER.debug("[ZBufferCapture] Wrote ${outputFile.name} (${data.width}x${data.height}, ${data.depthValues.size} floats)")
    }

    /**
     * Returns the output directory for Z-buffer captures.
     * ~/Documents/OysterClips/active_session/zbuffer/
     */
    private fun getOutputDirectory(): File {
        val userHome = System.getProperty("user.home")
        return File(userHome, "Documents/OysterClips/active_session/zbuffer")
    }

    /**
     * Returns the current pending write count (for testing/monitoring).
     */
    fun getPendingCount(): Int = pendingCount.get()

    /**
     * Returns whether Z-buffer capture is enabled.
     */
    fun isEnabled(): Boolean = System.getenv(ENV_VAR) == "1"

    /**
     * Resets the tick counter (for testing).
     */
    fun resetTickCounter() {
        tickCounter.set(0)
    }
}
